src/ocean/math/BinaryHistogram.d

```diff
@@ -0,0 +1,384 @@
+
+    An integer, binary logarithmic statistics counter
+
+    Params:
+        MaxPow2 = the maximum power of two that is tracked in the lower bins of
+            the counter. Any values higher than 1<<MaxPow2 will be aggregated in
+            the upper bin
+        Suffix = suffix for the names of the fields of the Bins accessor struct
+
+*******************************************************************************/
+
+public struct BinaryHistogram ( uint MaxPow2, istring Suffix = "" )
+{
+    import core.bitop: bsr;
+    import ocean.core.Traits : FieldName;
+
+    /// The number of bins. (The specified number of powers of two, plus one bin
+    /// for 0B and one for ≥1<<MaxPow2.)
+    private const NumBins = MaxPow2 + 2;
+
+    /// The aggregated total from all calls to `record()`.
+    public ulong total;
+
+    /// The total number of calls to `record()`.
+    public uint count;
+
+    /***************************************************************************
+
+        The bins of the byte count histogram. Given a number, the index i of the
+        bin to increment is calculated as follows:
+
+        - If 1 ≤ n and n < 2^MaxPow2: i = floor(log_2(n)) + 1,
+        - otherwise, if t < 1: i = 0,
+        - otherwise, t ≥ 2^MaxPow2: i = NumBins - 1.
+
+        For example, for MaxPow2 == 2, the following bins exist:
+        0: 0,
+        1:       1, 2:  2 -   3, 3:   4 -   7, 4:   8 -  15,  5:  16 -   31,
+        6: 32 - 63, 7: 64 - 127, 8: 128 - 255, 9: 256 - 511, 10: 512 - 1023,
+        11: 1ki -  (2ki-1), 12:  2ki -  (4ki-1), 13:  4ki -  (8ki-1),
+        14: 8ki - (16ki-1), 15: 16ki - (32ki-1), 16: 32ki - (64ki-1),
+        17: 64ki - ∞
+
+    ***************************************************************************/
+
+    private uint[NumBins] bins;
+
+    /***************************************************************************
+
+        Struct with one uint field per bin (see this.bins), named as follows
+        (assuming Suffix == ""):
+            from_0, from_1, from_2, from_4, from_8,
+            from_16, ...,
+            from_1Ki, from_2Ki, ...,
+            etc
+
+        Useful, for example, for logging the whole histogram. An instance of
+        this struct is returned by `stats()`.
+
+    ***************************************************************************/
+
+    private struct Bins
+    {
+        import ocean.core.Traits : ctfe_i2a;
+
+        /***********************************************************************
+
+            Interprets the passed bins array as a Bins instance.
+
+            Params:
+                array = bins array to reinterpret
+
+            Returns:
+                the passed bins array as a Bins instance
+
+        ***********************************************************************/
+
+        public static Bins fromArray ( typeof(BinaryHistogram.bins) array )
+        {
+            return *(cast(Bins*)array.ptr);
```

Comment:
It is not immediately obvious to me why do you need store the same data as both `Bins` struct and `ubyte` array.

Reply:
Good question. Originally, the bins array was also public. I made it private because I couldn't see a usage of it. You're right, though: the final step would be to remove it entirely.

Perhaps @david-eckardt-sociomantic has some insight into this, as the original author.